### PR TITLE
enum_shares: Cleanup and support non-meterpreter sessions

### PR DESCRIPTION
**Pull request by bcoles**
Resolves Rubocop violations.

Adds documentation.

Adds `Notes` module meta information.

Adds support for non-Meterpreter sessions.

### CODE REVIEW

File: modules/post/windows/gather/enum_shares.rb [hunk 1]
```diff
@@ -12,174 +12,199 @@ def initialize(info = {})
...
+    print_status('The following shares were found:')
+    share_names.each do |sname|
+      info = registry_getvaldata(shares_key, sname)
+      next if info.nil?
+
+      print_status("\tName: #{sname}")
+
+      # NOTE: Extracting these values will fail on Meterpreter.
+      # Share info keys are REG_MULTI_SZ. Meterpreter stdapi registry
+      # query_value() function returns malformed REG_MULTI_SZ values.
```
Comment (bcoles) [inline modules/post/windows/gather/enum_shares.rb:132]: This was broken both before and after this PR (https://github.com/rapid7/metasploit-payloads/issues/582).

```
#<Rex::Post::Meterpreter::Extensions::Stdapi::Sys::RegistrySubsystem::RegistryValue:0x000055e45e5363b0 @client=#<Session:meterpreter 192.168.200.218:51781 (192.168.200.218) "CORP\user @ WIN-17B09RRRJTG">, @hkey=0, @name="NETLOGON", @type=7, @data="C\x00S\x00C\x00F\x00l\x00a\x00g\x00s\x00=\x002\x005\x006\x00\x00\x00M\x00a\x00x\x00U\x00s\x00e\x00s\x00=\x004\x002\x009\x004\x009\x006\x007\x002\x009\x005\x00\x00\x00P\x00a\x00t\x00h\x00=\x00C\x00:\x00\\\x00W\x00i\x00n\x00d\x00o\x00w\x00s\x00\\\x00S\x00Y\x00S\x00V\x00O\x00L\x00\\\x00s\x00y\x00s\x00v\x00o\x00l\x00\\\x00c\x00o\x00r\x00p\x00.\x00l\x00o\x00c\x00a\x00l\x00\\\x00S\x00C\x00R\x00I\x00P\x00T\x00S\x00\x00\x00P\x00e\x00r\x00m\x00i\x00s\x00s\x00i\x00o\x00n\x00s\x00=\x000\x00\x00\x00R\x00e\x00m\x00a\x00r\x00k\x00=\x00L\x00o\x00g\x00o\x00n\x00 \x00s\x00e\x00r\x00v\x00e\x00r\x00 \x00s\x00h\x00a\x00r\x00e\x00 \x00\x00\x00S\x00h\x00a\x00r\x00e\x00N\x00a\x00m\x00e\x00=\x00N\x00E\x00T\x00L\x00O\x00G\x00O\x00N\x00\x00\x00T\x00y\x00p\x00e\x00=\x000\x00\x00\x00\x00\x00">
```

I believe Op3n4M3 may be looking into this, but should not be considered a blocker for this PR.

File: modules/post/windows/gather/enum_shares.rb [hunk 2]
```diff
@@ -12,174 +12,199 @@ def initialize(info = {})
...
-      keys = open_key.enum_key
-      keys.each do |k|
-        user_sid << k if k =~ /S-1-5-21-\d*-\d*-\d*-\d{3,6}$/
-      end
-      user_sid.each do |us|
-        mount_history = mount_history + enum_recent_mounts("HKU\\#{us.chomp}") if datastore["RECENT"]
-        run_history = run_history + enum_run_unc("HKU\\#{us.chomp}") if datastore["ENTERED"]
+      keys = registry_enumkeys('HKU') || []
+      keys.each do |maybe_sid|
+        next unless maybe_sid.starts_with?('S-1-5-21-')
```
Comment (smcintyre-r7) [inline modules/post/windows/gather/enum_shares.rb:186]: Would you mind defining the SID prefix here as a constant? We'll probably end up moving it somewhere eventually where we'll add additional well-known SIDs but for now just marking this one what it is.

`SID_PREFIX_USER`

Comment (bcoles) [reply modules/post/windows/gather/enum_shares.rb]: This seems like unnecessary and irrelevant scope creep for this PR but ok.

```# grep -rn S-1-5-2 modules/ lib/
# grep -rn S-1-5-2 modules/ lib/
modules/post/windows/gather/enum_shares.rb:186:        next unless maybe_sid.starts_with?('S-1-5-21-')
modules/post/windows/gather/credentials/nimbuzz.rb:41:      next unless k.include? "S-1-5-21"
modules/post/windows/gather/credentials/idm.rb:45:      next unless k.include? "S-1-5-21"
modules/post/windows/gather/enum_logged_on_users.rb:87:      next unless maybe_sid.starts_with?('S-1-5-21-')
modules/post/windows/manage/ie_proxypac.rb:118:      next unless k.include? "S-1-5-21"
modules/post/windows/manage/ie_proxypac.rb:151:      next unless k.include? "S-1-5-21"
modules/post/windows/manage/ie_proxypac.rb:173:      next unless k.include? "S-1-5-21"
modules/auxiliary/admin/kerberos/ms14_068_kerberos_checksum.rb:46:        OptString.new('USER_SID', [ true, 'The Domain User SID, Ex: S-1-5-21-1755879683-3641577184-3486455962-1000'])
modules/auxiliary/admin/kerberos/ms14_068_kerberos_checksum.rb:54:      print_error("Invalid USER_SID. Ex: S-1-5-21-1755879683-3641577184-3486455962-1000")
lib/msf/core/exploit/remote/kerberos/client/pac.rb:36:            # @option opts [String] :domain_id the domain SID Ex: S-1-5-21-1755879683-3641577184-3486455962
lib/msf/core/post/windows/user_profiles.rb:116:        next unless profkey.starts_with?('S-1-5-21')
```

**Review by smcintyre-r7 — APPROVED** [review]
Things look good to me now. I had to fix a minor issue with the SID constant definition. The discrepancy in the output between Meterpreter and the others is due to the documented issue that is unrelated to the proposed changes. I'll get this landed in a moment. Thanks @bcoles 

<details>
<summary>Testing Output</summary>

```
msf6 post(windows/gather/enum_shares) > sessions

Active sessions
===============

  Id  Name  Type                     Information                                                      Connection
  --  ----  ----                     -----------                                                      ----------
  2         shell x86/windows        Shell Banner: Microsoft Windows [Version 10.0.17763.3232] -----  192.168.159.128:4444 -> 192.168.159.10:57372 (192.168.159.10)
  3         powershell windows       DC$ @ DC                                                         192.168.159.128:4444 -> 192.168.159.10:57373 (192.168.159.10)
  4         meterpreter x64/windows  NT AUTHORITY\SYSTEM @ DC                                         192.168.159.128:4444 -> 192.168.159.10:57375 (192.168.159.10)

msf6 post(windows/gather/enum_shares) > run SESSION=2

[*] Running module against DC (192.168.159.10)
[*] The following shares were found:
[*] 	Name: SYSVOL
[*] 	Path: C:\Windows\SYSVOL\sysvol
[*] 	Remark: Logon server share 
[*] 	Type: DISK
[*] 
[*] 	Name: NETLOGON
[*] 	Path: C:\Windows\SYSVOL\sysvol\msflab.local\SCRIPTS
[*] 	Remark: Logon server share 
[*] 	Type: DISK
[*] 
[*] Post module execution completed
msf6 post(windows/gather/enum_shares) > run SESSION=3

[*] Running module against DC (192.168.159.10)
[*] The following shares were found:
[*] 	Name: SYSVOL
[*] 	Path: C:\Windows\SYSVOL\sysvol
[*] 	Remark: Logon server share 
[*] 	Type: DISK
[*] 
[*] 	Name: NETLOGON
[*] 	Path: C:\Windows\SYSVOL\sysvol\msflab.local\SCRIPTS
[*] 	Remark: Logon server share 
[*] 	Type: DISK
[*] 
[*] Post module execution completed
msf6 post(windows/gather/enum_shares) > run SESSION=4

[*] Running module against DC (192.168.159.10)
[*] The following shares were found:
[*] 	Name: SYSVOL
[*] 
[*] 	Name: NETLOGON
[*] 
[*] Post module execution completed
msf6 post(windows/gather/enum_shares) >
```
</details>

**Comment by smcintyre-r7** [comment]
# Release Notes
This cleans up the enum_shares post modules and adds support for shell sessions.

**Pull request merged**